src/pose_estimating.h

```diff
@@ -0,0 +1,229 @@
+ */
+struct Pose {
+  ConnectedJoint head;
+  ConnectedJoint neck;
+  ConnectedJoint shoulder;
+  ConnectedJoint hip;
+  ConnectedJoint knee;
+  ConnectedJoint foot;
+}
+
+/*
+ * @brief This class handles representations of the user's pose and calculates any updates to their pose that is required.
+ * 
+ * This class stores representations of the users current pose and their calibrated ideal pose, from the results of the PostProcessor class.
+ *
+ * This class handles:
+ * Updating current user's pose or ideal user's pose
+ * Calculating the angle between ConnectedJoints
+ * Comparing the current user's pose and ideal user's pose
+ *
+ * General use should just consist of updateAndCheckCurrentPose()
+ * 
+ */
+class PoseEstimater {
+ private:
+  /*
+   * @brief Calculates the angle(in degrees) between two points, anticlockwise from origin / bottom left corner of image
+   * @param x1 Relative X co-ordinate of first point
+   * @param y1 Relative Y co-ordinate of second point
+   * @param x2 Relative X co-ordinate of first point
+   * @param y2 Relative Y co-ordinate of first point
+   *
+   * @returns float
+   */
+  float getLineAngle(float x1, float y1, float x2, float y2)
+
+  /*
+   * @brief Converts `PostProcesser::ProcessedResults` to a `Pose::Pose`
+   *
+   * @param results  The PostProcessor::ProccessedResults struct from PostProcessor::PostProcessor being run
+   * @return JointsPose
+   */
+  Pose createPose(PostProcessing :: ProcessedResults results);
+
+  /*
+   * @brief sets user's current pose
+   *
+   * @param Pose
+   */
+  void set_current_pose(Pose current);
+
+  /*
+   * @brief sets user's ideal pose
+   *
+   * @param Pose
+   */
+  void set_ideal_pose(Pose current);
+
+  /*
+   * @brief sets calcaluted changes needed pose
+   *
+   * @param Pose
+   */
+  void set_pose_changes_needed(Pose current);
+
+ public:
+  /*
+   * @brief The user's current pose from most recent data provided
+   */
+  Pose current_pose;
+
+  /* 
+   * @brief The user's ideal pose, expected to be generated during user calibration  
+   */
+  Pose ideal_pose;
+
+  /*
+   * @brief A representation of what changes are needed to get back to ideal pose
+   */
+  Pose pose_changes_needed;
+
+  /*
+   * @brief A representation of what absolute value changes for a pose is acceptably still a good pose
+   */
+  float pose_change_threshold;
+
+  /*
+   * @brief Whether the user is in a good pose (true) or bad pose (false)
+   */
+  bool good_pose;
+
+  /*
+   * @brief Returns the user's current pose
+   * 
+   * @returns Pose
+   */
+  Pose get_current_pose();
+
+  /*
+   * @brief Updates the user's current pose from PostProcesser results
+   *
+   * @param results ProcessedResults struct containing user's pose data
+   */
+  void update_current_pose(PostProcessing::ProcessedResults results);
+
+  /*
+   * @brief Returns the user's ideal pose
+   * 
+   * @returns Pose
+   */
+  Pose get_ideal_pose();
```

Comment:
Would this be any different than accessing the `ideal_pose` field?

Reply:
> Some final remarks from me. I think what I'm getting at with my comments is that I'm not sure where the calculations are done: in the getters or setters? To my mind the options are:
> 
> **Update in setter**
> 
>     1. Call `update_pose()`
> 
>     2. All calculations are done in the setter and written to public fields
> 
>     3. The next block can read from the public fields
> 
> 
> **Update in getter**
> 
>     1. Call `update_pose()`
> 
>     2. Convert everything to an internal representation, but don't calculate everything yet
> 
>     3. Call `get_output()`
> 
>     4. This triggers the calculations and returns the desired output
> 
> 
> Basically it seems a bit confusing to have getters/setters and public member variables for the class.
> 
> On a separate note I'm not sure what the implications are here for multi-threading. If one thread updates the real posture, but another one tries to access get the interpreted pose there will be a race condition.
> 
> That's why I suggested a `run()` method. We can call it something else like:
> 
> ```c++
> Pose get_pose_changes_needed_based_on_current(ProcessedResults results)
> ```
> 
> As far as I can tell we will never call `update_current_pose()` without following with `get_pose_changes_needed()` or `check_good_posture()` so we should probably just allow either of these methods by providing only two public functions:
> 
>     * `updateCurrentPoseAndCheck()`
> 
>     * `updateCurrentPoseAndGetPoseChangesNeeded()`
> 
> 
> right?

That all makes sense. I just feel to have a complete class, the other functions should be there, even if we only ever use the run() method